Add few more e2e tests, with browser-mocked Jetpack and Google connection

tomalec opened the pull request:
### Changes proposed in this Pull Request:

This is a baby step towards #886.

- Add tests for a part of step 1 of MC onboarding. (2878bcd)
	```
  At setup page
    Merchant who is getting started
      ✓ should see accounts step header, "Connect your WordPress.com account" & connect button (1325ms)
      after clicking the "Connect your WordPress.com account" button
        ✓ should sent an API request to connect Jetpack, and redirect to the returned URL (2295ms)
    Merchant who has their Jetpack connected, but Google not yet connected
      ✓ should see their WPORG email, "Connect your Google account" text & connect button (1226ms)
      after clicking the "Connect your Google account" button
        ✓ should sent an API request to connect Jetpack, and redirect to the returned URL (2302ms)
	```
	Mock Jetpack and Google connection requests on the browser level. Later we could improve it by mocking on PHP/WCS level.

	Shorten the landing page tests. The details are now tested in `setup-mc/` tests.


- Add improved version of `createURL`. (e800471)
	as the `@wordpress/e2e-utils`/`@woocommerce/e2e-utils` is not helpful for absolute paths from `@woocommerce/e2e-utils` and merging `searchParams` which we'd need everytime.

	We should keep it until WordPress/gutenberg#34051 is merged or WordPress/gutenberg#34044 otherwise fixed.
- Add an improved version of `visitAdminPage`. (549f938)
	as the `@wordpress/e2e-utils`/`@woocommerce/e2e-utils` one is not helpful for absolute paths from `@woocommerce/e2e-utils` and merging `searchParams` which we'd need everytime.

	Could be simplified after WordPress/gutenberg#34051 is merged or WordPress/gutenberg#34044 otherwise fixed.

- Add a helper to mock XHR requests in e2e tests. (74f7b3d)
	Wrap Puppeteer's `page.setRequestInterception` and `page.on` with some sugar, to simplify mocking individual URLs.




### Screenshots:
![Animations of tests being run in browser](https://user-images.githubusercontent.com/17435/130120500-50eff190-f1f4-43cd-bb46-f20a65dcdf4f.gif)



### Detailed test instructions:

1. Checkout this branch, 
2. `npm run docker:up`
3. `npm run test:e2e`
4. `npm run test:e2e-dev` - to see what's happening, and to test a slightly different browser setup :confused: 

### Additional notes:

1. There are still plenty of tests to be written
2. This PR mocks Jetpack nad Google request on a browser level, it has a number of downsides:
	- Does not test -to-end
	- It requires even more mocking, as the server-side still lacks the connected state, so the following requests may fail
	
	But this is at least a small step forward, so we could:
	- write more tests, to avoid purely JS-based bugs, like https://github.com/woocommerce/google-listings-and-ads/issues/884
	- iteratively mock deeper, add similar helpers to mock WCS, etc.
3. I believe, that even once we will mock Jetpack and Google calls deeper, the `RequestMock` util could be helpful for spying on those requests and responses.

<!--
Optional.
Enter a summary of all changes in this Pull Request, which will be added to the changelog if accepted.
Each line should start with `(Fix|Add|Tweak|Update) - `, for example:
> Fix - I took care of something that wasn't working.
> Add - I added something new that's pretty cool.
> Tweak - I made a small change.
> Update - I made big changes to something that wasn't broken.

Leave the "Changelog entry" header in place completely empty, without any summary if no changelog entry is needed.
If you remove the "Changelog entry" header, the title of Pull Request will be used as the changelog entry.
-->
### Changelog entry

> Tweak - Added a few more e2e tests and utils.

tomalec committed:
Add improved version of `createURL`.

as the `@wordpress/e2e-utils`/`@woocommerce/e2e-utils` is not helpful for absolute paths from `@woocommerce/e2e-utils` and merging `searchParams` which we'd need everytime.

We should keep it until https://github.com/WordPress/gutenberg/pull/34051 is merged or https://github.com/WordPress/gutenberg/issues/34044 otherwise fixed.

tomalec committed:
Add improved version of `visitAdminPage`.

as the `@wordpress/e2e-utils`/`@woocommerce/e2e-utils` one is not helpful for absolute paths from `@woocommerce/e2e-utils` and merging `searchParams` which we'd need everytime.

Could be simplified after https://github.com/WordPress/gutenberg/pull/34051 is merged or https://github.com/WordPress/gutenberg/issues/34044 otherwise fixed.

tomalec committed:
Add a helper to mock XHR requests in e2e tests.

Wrap Puppeteer's `page.setRequestInterception` and `page.on` with some sugar, to simplify mocking individual URLs.

tomalec committed:
Add tests for a part of step 1 of MC onboarding.

Mock Jetpack and Google connection requests on browser level. Later we could improve it by mocking on PHP/WCS level.

Shorten the landing page tests. The details are now tested in `setup-mc/` tests.

This is a baby-step towards https://github.com/woocommerce/google-listings-and-ads/issues/886.

findingsimple commented:
🌶️ 🔥 nice work @tomalec

eason9487 commented on tests/e2e/utils/create-url.js:
````diff
@@ -0,0 +1,42 @@
+ * // "http://localhost:8889/foo?a=b&bar=baz&fiz=a%2Fb%2Fc"
+ * ```
+ *
+ * @param {string}              [WPPath=''] String to be serialized as pathname.
+ * @param {string|Object|Array} [query]     Query parameters in any format supported by {@link https://developer.mozilla.org/en-US/docs/Web/API/URLSearchParams/URLSearchParams | URLSearchParams}.
+ * @return {string} String which represents full URL.
+ */
+export function createURL( WPPath = '', query = '' ) {
+	const url = new URL( WPPath, WP_BASE_URL );
````
❓ 
The `WPPath` is passed to `new URL` as the first parameter, so it could accept absolute or relative URL. Consider that the meaning of this variable has been changed, I think the variable name and its JSDoc might need to be adjusted as well.

tomalec replied:
Initially, I wanted to limit the number of changes to the minimum in the upstream PR, but you're with, the changed behavior should be reflected in the name.

I wonder what would be better. Formally that's "url string": https://url.spec.whatwg.org/#url-writing : relative or absolute URL string with almost everything optional.
- `url` - to match MDN definition of `new URL( url )`, and somehow match `url string` in spec, but give a premise of being not a string but something stringifiable to URL string?

Then we could have something like:
Creates new absolute (full) URL string by parsing given url and `query` relative to the WP base.
```js
/**
 * Creates new absolute (full) URL string by parsing given
 * `url` and `query` relative to the WP base.
 *
 * @param {string}              [url=''] A valid URL string, usually a relative path, but could contain as well contain query, fragment or be absolute.
 * @param {string|Object|Array} [query]  Query parameters in any format supported by {@link https://developer.mozilla.org/en-US/docs/Web/API/URLSearchParams/URLSearchParams | URLSearchParams}.
 * @return {string} String which represents the full, absolute URL.
 */
export function createURL( url = '', query = '' ) {
	const fullUrl = new URL( url, WP_BASE_URL );
```
 
WDYT?

eason9487 replied:
I think it's good, and in this way, it can match the absolute URL usage of the below as well. 👍 

https://github.com/woocommerce/google-listings-and-ads/blob/2878bcd89b9cbfc122c74e230ba1ac5873f03e7d/tests/e2e/utils/visit-gla-page.js#L30

eason9487 commented on tests/e2e/utils/request-mock.js:
```diff
@@ -0,0 +1,103 @@
+	 *
+	 * @param {import("puppeteer").Request} interceptedRequest
+	 */
+	#callMocks( interceptedRequest ) {
```
❔ 
I tested e2e test by replacing the line with `#callMocksOnThis = ( interceptedRequest ) => {`, and it works. Is there any concern about not adopting this syntax?

tomalec replied:
yes, as then the same function will be shared between all instances when it's added/removed by the event listener. So If one instance will remove its listener, it will make all other instances stop listening.

tomalec replied:
If the puppeteer API would support [`EventListener`s](https://developer.mozilla.org/en-US/docs/Web/API/EventListener) in `page.addEventListener`, we could use it like:
```js
	async observe( page ) {
		// Unobserve the previous page.
		this.disconnect();

		this.#observedPage = page;
		await page.setRequestInterception( true );
		page.addEventListener( 'request', this );
	}
	disconnect() {
		if ( this.#observedPage ) {
			this.#observedPage.removeEventListener( 'request', this );
			this.#observedPage = undefined;
		}
	}

	/**
	 * Call a matching mock, if any.
	 *
	 * @param {import("puppeteer").Request} interceptedRequest
	 */
	#handleEvent( interceptedRequest ) {
```
without the need for binding an instance of a function to the `RequestMock` instance.

Like for any other Web listeners https://developer.mozilla.org/en-US/docs/Web/API/EventTarget/addEventListener



-----

But it seems for puppeteer `addEventListener` is notto add and `EventListener` but to `addCallback` ;) :see_no_evil:

eason9487 replied:
Sorry for not being clear, what I meant was the method syntax `method = () => {}` is still binding `this` on the function itself, so it would not be a prototype method. In this way, it could make it no need to bind the prototype method `callMocks` in the constructor and an additional member `callMocksOnThis`. I think it should be equivalent to the current implementation.

For example:

```js
class Foo {
	msg = 'Hi';
	bar = () => console.log( this.msg );
}
```

It would not be shared between all instances and could be invoked as a normal function call directly.

![image](https://user-images.githubusercontent.com/17420811/130557225-7baee06b-5f42-45b2-b29f-3feffad28f0c.png)

tomalec replied:
ah, ok. So you mean remove it from class/prototype and add to the instance only?

Sure it will work but will be slightly less extendable, as we will have no function to inherit by things like `WCSMock`, `GoogleAPIMock`, etc. that was thinking we could have one day.
But as they are not here today, and I still struggle with bigger problems to implement them, maybe we can simplify the current shape for the sake of YAGNI, or TheyAintGonnaByUsefull ;)

eason9487 replied:
> Sure it will work but will be slightly less extendable, as we will have no function to inherit by things like `WCSMock`, `GoogleAPIMock`, etc. that was thinking we could have one day.

Since I rarely use JavaScript inheritance in my practical experience except for exercise, I don't have a concrete opinion about that. I was just curious why you chose to use the current implementation. I'm good with the current way.

tomalec replied:
I changed it in [`dabc336`](https://github.com/woocommerce/google-listings-and-ads/pull/970/commits/dabc33650c2482a0d2bd0ab6cc7f14cf551d4966) to 
```js
#callMock;
constructor() {
	this.#mocks = new Map();
	/**
	 * Calls a matching mock, if any.
	 *
	 * @param {import("puppeteer").Request} interceptedRequest
	 */
	this.#callMock = ( interceptedRequest ) => {
```
Unfortunately, VSCode's JSDoc fails to pick up private instance methods, leaving absolutely no hint for `#callMock`property which is assigned to a function statement.

We can change it to `_callMock`, to get IDE support, as private methods are not well addopted anyway.

eason9487 commented on tests/e2e/specs/setup-mc/step-1-accounts.test.js:
```diff
@@ -0,0 +1,187 @@
+
+		describe( 'after clicking the "Connect your Google account" button', () => {
+			let googleConnectMock;
+			beforeEach( async function mockJetpackConnect() {
```
💅 
The function `mockJetpackConnect` looks like missed renaming? 🧐

eason9487 commented on tests/e2e/specs/setup-mc/step-1-accounts.test.js:
```diff
@@ -0,0 +1,187 @@
+					} );
+			} );
+
+			it( 'should sent an API request to connect Jetpack, and redirect to the returned URL', async () => {
```
💅 
The `"Jetpack"` looks like missed replacing to a text such as "Google auth".

eason9487 commented on tests/e2e/specs/setup-mc/step-1-accounts.test.js:
```diff
@@ -0,0 +1,187 @@
+			} );
+
+			it( 'should sent an API request to connect Jetpack, and redirect to the returned URL', async () => {
+				const connectWPButton = await page.waitForXPath(
```
💅 
Would it be more proper to name it as something like `connectGoogleButton`?

tomalec committed:
Rename `createURL` parameter, update JSDoc,

to make the name reflect closer the behavior.

Addresses https://github.com/woocommerce/google-listings-and-ads/pull/970#discussion_r693844751.

tomalec committed:
Update variables, functions and descriptions used in e2e tests,

to match what they actually are.
Addresses https://github.com/woocommerce/google-listings-and-ads/pull/970#pullrequestreview-737102056.

tomalec committed:
Remove `callMocks` from class andd it to an instance.

Addresses https://github.com/woocommerce/google-listings-and-ads/pull/970#discussion_r693859320.

tomalec commented:
I believe I addressed all the comments, and the PR is ready for another review.

eason9487 reviewed (approved):
Ran e2e test on my env and works well. LGTM.
